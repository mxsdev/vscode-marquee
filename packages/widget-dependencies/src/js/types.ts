import type vscode from 'vscode'
import type { DependencyType } from 'src/types'

export type JsPackageManager = 'npm'|'yarn'|'pnpm'

export interface JsPackageDependency {
  name: string
  dependencyType?: DependencyType
  /**
   * Query version in package.json
   */
  packageVersion: string
  /**
   * Version in node_modules
   */
  actualVersion?: string
  homepage?: string
  repository?: string
}

export interface LocalJsPackage extends JsPackageMeta {
  isWorkspace?: boolean
  name: string
  version: string
  dependencies: Record<string, JsPackageDependency>
  uri: vscode.Uri
  workspaces?: string[]
  manager: JsPackageManager
}

export type LocalJsProjects = Record<string, LocalJsPackage>

export type RemoteRegistry = Record<string, JsPackageMeta>

export interface JsPackageMeta {
  latestVersion?: string
  homepage?: string
  repository?: string
}

export interface NPMOutdatedInfo {
  packageId: string
  current?: string
  wanted?: string
  latest?: string
  workspace?: string
  packageType?: 'devDependencies'|'dependencies'
  url?: string
}

export type NPMOutdatedHead = (
  | 'Package'
  | 'Current'
  | 'Wanted'
  | 'Latest'
  | 'Workspace'
  | 'Package Type'
  | 'URL'
)[]

export type NPMOutdatedColumn = NPMOutdatedHead[number]

export interface NPMOutdatedData {
  body: string[][]
  head: NPMOutdatedHead
}