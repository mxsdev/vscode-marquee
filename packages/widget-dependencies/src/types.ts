import type { ContextProperties } from '@vscode-marquee/utils'
import type vscode from 'vscode'
export type { RawPackument } from 'query-registry'

export interface State {
  dependencies: DisplayedDependency[]
  loading: boolean
  capabilities: DependencyProviderCapabilities
  event: (EventsObj & { id: EventId }) | {}
}

export interface Configuration {
  /* General */
  showUpToDate: boolean
  autoRefresh: boolean

  /* JS */
  jsRegistry: string
  prefersPnpm: boolean
}

export interface Events {
  refreshDependencies: {}
  updateDependency: { packageId: string, toVersion: string, workspace?: string }
  updateAllDependencies: {}
  removeDependency: { packageId: string, workspace?: string, isRootWorkspace: boolean }
}

export interface Context extends ContextProperties<Configuration & State> {
  _refreshDependencies(): void
  _updateDependency(dep: DisplayedDependency, toVersion: string): void
  _updateAllDependencies(): void
  _removeDependency(dep: DisplayedDependency): void
}

export interface DependencyProvider extends vscode.Disposable {
  loadDependencies(): Promise<DisplayedDependency[]|undefined>

  upgradeDependency?: (
    packageId: string,
    upgradeTo: string,
    workspace: string|undefined
  ) => Promise<void>

  upgradeAllDependencies?: () => Promise<void>

  deleteDependency?: (
    packageId: string,
    workspace: string|undefined,
    isRootWorkspace: boolean
  ) => Promise<void>

  /**
   * Whether or not this provider explicitly sets if this dependency needs
   * an upgrade via the `needsUpgrade` prop of `DisplayedDependency`
   *
   * If this is false, the client determines whether a package needs an update
   * by comparing the current and latest semantic versions
   */
  explicitNeedsUpgrade?: () => Promise<boolean>
}

export interface TerminalProvider extends vscode.Disposable {
  getOrCreateTerminal(cwd: vscode.Uri): vscode.Terminal
}

export type DependencyType = 'normal'|'dev'

export interface DisplayedDependency {
  name: string

  versions: {
    current?: string
    wanted?: string
    latest?: string
    query?: string
  },

  project?: string

  isRootWorkspace: boolean

  dependencyType?: DependencyType

  url?: string

  needsUpgrade?: boolean
}

export interface DependencyProviderCapabilities {
  deleteDependency: boolean
  upgradeDependency: boolean
  upgradeAllDependencies: boolean
  explicitNeedsUpgrade: boolean
}

export type EventsObj<K extends keyof Events = keyof Events>
  = K extends string ? { type: K, payload: Events[K] } : never

export type EventId = number
