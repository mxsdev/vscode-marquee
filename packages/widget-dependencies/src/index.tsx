import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faShareNodes } from '@fortawesome/free-solid-svg-icons/faShareNodes'

import DependenciesWidget from './Widget'

export default {
  name: 'dependencies',
  icon: <FontAwesomeIcon icon={faShareNodes} />,
  label: 'Dependencies',
  tags: ['workspace', 'projects', 'productivity'],
  description: 'Dependencies for your current project.',
  component: DependenciesWidget
}
