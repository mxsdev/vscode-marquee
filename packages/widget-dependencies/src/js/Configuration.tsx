import { FormControlLabel, Checkbox, TextField } from '@mui/material'
import React, { useContext } from 'react'
import DependencyContext from '../Context'

const Configuration = () => {
  const {
    prefersPnpm,
    setPrefersPnpm,
    jsRegistry,
    setJsRegistry
  } = useContext(DependencyContext)
  
  return (
    <>
      <FormControlLabel 
        control={
          <Checkbox 
            checked={prefersPnpm}
            onChange={(e) => setPrefersPnpm(e.target.checked)}
          />
        }
        label="Prefer pnpm"
      />

      <TextField 
        id="widget-dependencies-config-js-registry" 
        label="Registry" 
        variant="filled"

        value={jsRegistry}
        onChange={(e) => setJsRegistry(e.target.value)}

        placeholder="Default registry..."
      />
    </>
  )
}

export default Configuration