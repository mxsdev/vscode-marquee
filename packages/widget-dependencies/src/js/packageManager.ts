import type { PackageJSON } from 'query-registry'
import vscode  from 'vscode'
import type { 
  JsPackageManager, 
  LocalJsPackage, 
  LocalJsProjects, 
  NPMOutdatedColumn, 
  NPMOutdatedData, 
  NPMOutdatedInfo 
} from './types'

declare const IS_WEB_BUNDLE: boolean

export async function tryGetPackageJson (
  uri: vscode.Uri,
): Promise<PackageJSON|undefined> {
  const pkgJsonPath = vscode.Uri.joinPath(uri, 'package.json')
    
  let file: string

  try {
    const rawBytes = (await vscode.workspace.fs.readFile(pkgJsonPath))
    // buffer.toString() not supported on web
    file = IS_WEB_BUNDLE ? new TextDecoder().decode(rawBytes) : rawBytes.toString()
  } catch(e) {
    // no package.json found
    return undefined
  }

  let pkgJson: PackageJSON

  try {
    pkgJson = JSON.parse(file)
  } catch(e) {
    throw new Error('Unable to serialize package.json')
  }

  if (typeof pkgJson !== 'object' 
    || typeof pkgJson.name !== 'string'
    || typeof pkgJson.version !== 'string'
  ) { 
    throw new Error('Invalid package.json')
  }

  return pkgJson
}

export async function tryGetJsProject (
  uri: vscode.Uri,
  parent?: LocalJsPackage,
  prefersPnpm?: boolean,
): Promise<LocalJsPackage|undefined> {
  const pkgJson = await tryGetPackageJson(uri)
  if(!pkgJson) { return undefined }

  const repositoryUrl = typeof pkgJson.repository === 'string' && pkgJson.repository
  const workspaces = pkgJson.workspaces as string[]|{ projects: string[] }|undefined

  const pkg: LocalJsPackage = {
    name: pkgJson.name,
    version: pkgJson.version,
    dependencies: {},
    isWorkspace: !!parent,
    workspaces: Array.isArray(workspaces) ? workspaces : workspaces?.projects,
    uri,
    manager: parent?.manager ?? await getPackageManager(uri, prefersPnpm),
    homepage: pkgJson.homepage,
    repository: repositoryUrl || undefined
  }

  const dependenciesZipped = [
    ...Object.entries(pkgJson.dependencies ?? {}).map(pkg => ([pkg, 'normal'] as const)),
    ...Object.entries(pkgJson.devDependencies ?? {}).map(pkg => ([pkg, 'dev'] as const))
  ]

  await Promise.all(
    dependenciesZipped.map(async ([[name, version], dependencyType]) => {
      const localJson = await tryGetPackageJson(
        vscode.Uri.joinPath(pkg.uri, 'node_modules', ...name.split('/'))
      ) ?? (parent && await tryGetPackageJson(
        vscode.Uri.joinPath(parent.uri, 'node_modules', ...name.split('/'))
      ))

      const repo = typeof localJson?.repository === 'string' ? localJson?.repository : undefined

      pkg.dependencies[name] = {
        name,
        packageVersion: version,
        actualVersion: localJson?.version,
        dependencyType,
        homepage: localJson?.homepage,
        repository: repo
      }
    })
  )

  return pkg

  async function getPackageManager (
    uri: vscode.Uri,
    prefersPnpm?: boolean,
  ): Promise<JsPackageManager> {
    try {
      await vscode.workspace.fs.stat(vscode.Uri.joinPath(uri, 'yarn.lock'))
      return 'yarn'
    } catch(e) { }

    return prefersPnpm ? 'pnpm' : 'npm'
  }
}

export async function findWorkspaces (
  rootProject: LocalJsPackage,
): Promise<LocalJsProjects> {
  const projects: LocalJsProjects = {
    [rootProject.name]: rootProject,
  }

  // apply glob patterns in workspaces
  const candidateUris = (await Promise.all(
    rootProject.workspaces?.map(workspacePath => 
      vscode.workspace.findFiles(
        new vscode.RelativePattern(rootProject.uri, workspacePath)
      )
    ) ?? []
  )).flatMap(x => x)
  
  // filter for only the folders, and add to projects
  await Promise.all(
    candidateUris.map(async (relativeUri) => {
      const uri = vscode.Uri.joinPath(rootProject.uri, relativeUri.fsPath)

      const { type } = await vscode.workspace.fs.stat(uri)

      if(type !== vscode.FileType.Directory) { return }

      const pkg = await tryGetJsProject(uri, rootProject)
      if(!pkg) { return }

      projects[pkg.name] = pkg
    })
  )

  return projects
}

export async function packageManagerCmdOutdated (
  pkg: LocalJsPackage,
): Promise<NPMOutdatedInfo[]|undefined> {
  const output = await packageManagerCmd(
    pkg, 
    'outdated',
    '--json',
  )

  if(!output) { return undefined }

  const lines = output.split('\n')

  const table = lines
    .find((line) => {
      try {
        const json = JSON.parse(line)
        return json.type === 'table'
      } catch(e) {
        return false
      }
    })
  
  if(!table) { return undefined }

  let data: NPMOutdatedData

  try {
    data = JSON.parse(table).data
  } catch(e) {
    return undefined
  }

  // invert header
  const idxMap = data.head.reduce(
    (prev, col, i) => {
      prev[col] = i
      return prev
    }, 
    {} as Record<NPMOutdatedColumn, number>
  )

  // Need packageId
  if(!('Package' in idxMap)) {
    return undefined
  }

  const fromRow = (from: string[], col: NPMOutdatedColumn) => {
    if (col in idxMap) { 
      return from[idxMap[col]] 
    }

    return undefined
  }

  return data.body.map(row => ({
    packageId: fromRow(row, 'Package')!,
    current: fromRow(row, 'Current'),
    latest: fromRow(row, 'Latest'),
    packageType: fromRow(row, 'Package Type') as NPMOutdatedInfo['packageType'],
    url: fromRow(row, 'URL'),
    wanted: fromRow(row, 'Wanted'),
    workspace: fromRow(row, 'Workspace')
  }))
}

export async function packageManagerCmd (
  pkg: LocalJsPackage,
  ...args: string[]
): Promise<string|undefined> {
  if(IS_WEB_BUNDLE) { return undefined }

  const { exec } = require('child_process') as typeof import('node:child_process')
  
  return await new Promise<string|undefined>((res, rej) => {
    const proc = exec(
      [pkg.manager, ...args].join(' '),
      { cwd: pkg.uri.fsPath },
      (err,stdout,stderr) => {
        if(stdout) {
          res(stdout)
        }

        if(err) {
          rej(err)
        }

        if(stderr) {
          rej(stderr)
        }
      }
    )

    proc.on('close', () => res(undefined))
  })
}