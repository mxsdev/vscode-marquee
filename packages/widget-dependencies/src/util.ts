import vscode from 'vscode'
import type { TerminalProvider } from './types'

declare const IS_WEB_BUNDLE: boolean

// https://decipher.dev/30-seconds-of-typescript/docs/URLJoin/
export function urlJoin (...args: string[]) {
  return args
    .join('/')
    .replace(/[\/]+/g, '/')
    .replace(/^(.+):\//, '$1://')
    .replace(/^file:/, 'file:/')
    .replace(/\/(\?|&|#[^!])/g, '$1')
    .replace(/\?/g, '&')
    .replace('&', '?')
}

export function getFetch (): typeof fetch {
  return IS_WEB_BUNDLE ? fetch : require('node-fetch').default
}

export class NamedTerminalProvider implements TerminalProvider {
  private terminal?: vscode.Terminal
  
  constructor (
    private name: string
  ) { }

  getOrCreateTerminal (): vscode.Terminal {
    if(this.terminal && !this.terminal.exitStatus) { return this.terminal }

    this.terminal?.dispose()
    
    this.terminal = vscode.window.createTerminal(
      this.name
    )

    return this.terminal
  }

  dispose () {
    this.terminal?.dispose()
  }
}