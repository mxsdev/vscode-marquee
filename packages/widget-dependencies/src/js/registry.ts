import { getFetch, urlJoin } from '../util'
import { RawPackument } from '../types'
import { DEFAULT_JS_REGISTRY } from '../constants'
import { LocalJsProjects, RemoteRegistry } from './types'
import { Logger } from '@vscode-marquee/utils/build/logger'

/**
 * Go through project/workspaces and pull package metadata for each dependency
 * from npm/remote registry provided by `registryUrl`
 */
export async function pullRemoteRegistry (
  projects: LocalJsProjects,
  registryUrl?: string
): Promise<RemoteRegistry> {
  const registry: RemoteRegistry = {}
  
  await Promise.all(
    Object.values(projects)
      .flatMap(repo => Object.values(repo.dependencies).map(async dep => {
        if(!projects[dep.name]) {
          const packument = await getRawPackument(dep.name, registryUrl)
          if(!packument) { return } // package does not exist
          
          const { versions, homepage, repository } = packument

          const version = packument['dist-tags'].latest

          const latest = versions[version]
          if(!latest) { return } // package has no releases

          const repositoryUrl = typeof repository === 'string' && repository

          registry[dep.name] = { 
            latestVersion: latest.version,
            homepage,
            repository: repositoryUrl || undefined
          }
        } else {
          registry[dep.name] = projects[dep.name]
        }
      }))
  )

  return registry
}

export async function getRawPackument (
  packageId: string, 
  registry: string|undefined
): Promise<RawPackument|undefined> {
  registry ||= DEFAULT_JS_REGISTRY
  const url = urlJoin(registry, packageId)

  const fetch = getFetch()

  return await fetch(url)
    .then(res => res.json() as Promise<RawPackument>)
    .catch((e) => { 
      Logger.error('Failed to fetch package', e)
      return undefined
    })
}