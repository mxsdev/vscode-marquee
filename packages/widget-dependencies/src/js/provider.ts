import vscode from 'vscode'
import { Configuration, DependencyProvider, DisplayedDependency, TerminalProvider } from 'src/types'
import { pullRemoteRegistry } from './registry'
import { findWorkspaces, packageManagerCmdOutdated, tryGetJsProject } from './packageManager'
import { JsPackageManager, LocalJsPackage } from './types'

declare const IS_WEB_BUNDLE: boolean

export class JsDependencyProvider implements DependencyProvider {
  private packageManager: JsPackageManager = 'npm'
  private disposables: vscode.Disposable[] = []
  
  constructor (
    private readonly config: Configuration,
    private terminalProvider: TerminalProvider,
    refreshDependencies: () => void,
  ) { 
    const watcher = vscode.workspace.createFileSystemWatcher(
      '**/package.json'
    )
    
    this.disposables.push(
      watcher
    )

    this.disposables.push(
      watcher.onDidChange(() => {
        refreshDependencies()
      })
    )
  }

  async loadDependencies (): Promise<DisplayedDependency[]|undefined> {
    const pkg = await this.getPackageInfo()
    if(!pkg) { return undefined }
  
    const projects = await findWorkspaces(pkg)

    if (!IS_WEB_BUNDLE) {
      const data = await packageManagerCmdOutdated(pkg) ?? []

      const upgradeableDepMemo: Record<string, Set<string>|undefined> = {}
      
      const upgradeableDeps: DisplayedDependency[] = await Promise.all(data.map(
        async ({
          packageId,
          current,
          wanted,
          latest,
          workspace,
          packageType,
          url
        }) => {
          const workspaceId = workspace ?? pkg.name
          
          if(!upgradeableDepMemo[workspaceId]) {
            upgradeableDepMemo[workspaceId] = new Set()
          }
          
          upgradeableDepMemo[workspaceId]?.add(packageId)

          const hasWorkspaces = (pkg.workspaces?.length ?? 0) > 0

          return ({
            name: packageId,
            versions: {
              current,
              wanted,
              latest,
              query: pkg.dependencies[packageId]?.packageVersion
            },
            dependencyType: packageType === 'dependencies' ? 'normal'
              : packageType === 'devDependencies' ? 'dev'
                : undefined,
            project: workspace,
            url,
            needsUpgrade: true,
            isRootWorkspace: hasWorkspaces && (!workspace || workspace === pkg.name)
          })
        }
      ))

      const upToDateDeps = this.config.showUpToDate && Object.values(projects)
        .flatMap(repo => 
          Object.values(repo.dependencies)
            .filter(dep => !upgradeableDepMemo[repo.name]?.has(dep.name))
            .map(dep => {
              const hasWorkspaces = (repo.workspaces?.length ?? 0) > 0

              return ({
                name: dep.name,
                versions: {
                  current: dep.actualVersion,
                  query: dep.packageVersion,
                  latest: dep.actualVersion,
                },
                ...repo.isWorkspace && { project: repo.name },
                dependencyType: dep.dependencyType,
                url: dep.homepage ?? dep.repository,
                isRootWorkspace: hasWorkspaces
              })
            })
        )

      return [
        ...upgradeableDeps,
        ...upToDateDeps || []
      ]
    } else {
      const remoteRegistry = await pullRemoteRegistry(projects, this.config.jsRegistry)

      return Object.values(projects)
        .flatMap(
          repo => 
            Object.values(repo.dependencies)
              .filter(dep => dep.name in remoteRegistry)
              .map(dep => {
                const remoteDep = remoteRegistry[dep.name]!

                const hasWorkspaces = (repo.workspaces?.length ?? 0) > 0

                return ({
                  name: dep.name,
                  versions: {
                    current: dep.actualVersion,
                    query: dep.packageVersion,
                    latest: remoteDep.latestVersion ?? undefined,
                  },
                  ...repo.isWorkspace && { project: repo.name },
                  dependencyType: dep.dependencyType,
                  url: remoteDep.homepage ?? remoteDep.repository,
                  isRootWorkspace: hasWorkspaces
                })
              })
        )
    }
  }

  async upgradeDependency (
    packageId: string, 
    toVersion: string, 
    workspace: string|undefined
  ) {
    this.packageManagerCommand(
      ['upgrade', `${packageId}@${toVersion}`],
      workspace
    )
  }

  async upgradeAllDependencies () {
    this.packageManagerCommand(['upgrade'])
  }

  async deleteDependency (
    packageId: string, 
    workspace: string|undefined,
    isRootWorkspace: boolean = false
  ) {
    this.packageManagerCommand(
      ['remove', packageId],
      workspace,
      isRootWorkspace
    )
  }

  async explicitNeedsUpgrade () {
    return !IS_WEB_BUNDLE
  }

  async getPackageInfo (): Promise<LocalJsPackage|undefined> {
    const { workspaceFolders } = vscode.workspace
    
    if(!workspaceFolders || workspaceFolders.length === 0) { return undefined }
    
    const workspacePath = workspaceFolders[0].uri
    const pkg = await tryGetJsProject(workspacePath, undefined, this.config.prefersPnpm)

    if(!pkg) { return undefined }

    this.packageManager = pkg.manager

    return pkg
  }

  packageManagerCommand (
    args: string[],
    workspace?: string,
    rootWorkspaceFlag = false
  ) {
    args = [
      ...rootWorkspaceFlag ? ['-W'] : [],
      ...(workspace && !rootWorkspaceFlag) ? ['workspace', workspace] : [],
      ...args
    ]
    
    const terminal = this.terminalProvider.getOrCreateTerminal()

    terminal.sendText(
      [ this.packageManager, ...args ].join(' '),
      true
    )

    terminal.show(true)
  }

  dispose () { 
    this.disposables.forEach(d => d.dispose())
  }
}