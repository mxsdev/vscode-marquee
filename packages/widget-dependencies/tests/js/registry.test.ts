import { getRawPackument, pullRemoteRegistry } from '../../src/js/registry'
import * as util from '../../src/util'

const fetch = jest.fn(
  async () => ({
    json: async () => ({
      homepage: 'https://homepage.com',
      repository: 'https://github.com/repo',
      'dist-tags': { latest: '0.1.0' },
      versions: {
        '0.1.0': {
          version: '0.1.0',
        }
      }
    })
  })
)

;(util as any).getFetch = () => fetch

afterEach(() => fetch.mockClear())

test('getRawPackument', async () => {
  const packument = await getRawPackument('package', 'https://registry.com')

  expect(fetch).toBeCalledWith('https://registry.com/package')

  expect(packument).toMatchSnapshot()
})

test('pullRemoteRegistry', async () => {
  const remoteRegistry = await pullRemoteRegistry (
    {
      repo: {
        dependencies: {
          foo: { name: 'foo', packageVersion: '0.0.1' },
          bar: { name: 'bar', packageVersion: '0.0.1' },
        }
      } as any
    },
    'https://registry.com'
  )

  expect(fetch).toBeCalledWith('https://registry.com/foo')
  expect(fetch).toBeCalledWith('https://registry.com/bar')

  expect(remoteRegistry).toMatchSnapshot()
})