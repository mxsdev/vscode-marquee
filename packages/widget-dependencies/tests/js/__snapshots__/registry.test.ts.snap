// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRawPackument 1`] = `
{
  "dist-tags": {
    "latest": "0.1.0",
  },
  "homepage": "https://homepage.com",
  "repository": "https://github.com/repo",
  "versions": {
    "0.1.0": {
      "version": "0.1.0",
    },
  },
}
`;

exports[`pullRemoteRegistry 1`] = `
{
  "bar": {
    "homepage": "https://homepage.com",
    "latestVersion": "0.1.0",
    "repository": "https://github.com/repo",
  },
  "foo": {
    "homepage": "https://homepage.com",
    "latestVersion": "0.1.0",
    "repository": "https://github.com/repo",
  },
}
`;
